Assertions for Functional.Unions (Union<...> and Task<Union<T...>>)

Justin-Cooney opened the pull request:
An early draft of the union assertions feature. This PR adds a new solution for union assertions which contains fluent assertions for Unions with two types. The assertion classes added here will have to be duplicated for Unions with three or more types.

The added assertions in this PR will allow both `AdHoc` and `Defined` Unions to be validated using the following syntax: 
```C#
SomeUnion
    .Value()
    .Should()
    .BeOfUnionType<TSubType>()
    .AndValue
    .Should()
    .Be(SomeObject)
```

Justin-Cooney committed:
Initial draft for the Union assertions feature. Adds Union assertions for Unions with two types.

Justin-Cooney commented on src/Functional.Unions.FluentAssertions/Functional.Unions.FluentAssertions/UnionValueTypeAssertions.cs:
```diff
@@ -0,0 +1,71 @@
+		/// <typeparam name="TExpected">The type in the Union that the value is expected to be of.</typeparam>
+		/// <param name="because">Additional information for if the assertion fails.</param>
+		/// <param name="becauseArgs">Zero or more objects to format using the placeholders in <paramref name="because"/>.</param>
+		/// <param name="ignore">Ignore this parameter as it should always be the default value.</param>
```
This is one thing that has me dissatisfied. In order to differentiate each of the `BeOfUnionType` method overloads I've added this ignored parameter.

Obviously this isn't desirable, but I couldn't devise any other way to have a single method name check all possible types. Without the ignored parameter these methods would have to be `BeOfUnionTypeOne` and `BeOfUnionTypeTwo` which decreases usability.

Any thoughts on how acceptable this is or ideas for alternatives?

RyanMarcotte replied:
I wonder if you could instead have a single `BeOfUnionType<T>` method that does an ugly type-checking chain?  The intent is that the individual types in the union are different, right?

```
public AndUnionValueConstraint<T> BeOfUnionType<T>(...)
{
    // potentially some assertion here to ensure One and Two are different types?

    if (_subject.One() is Option<T> one)
        return AndUnionValueConstraint<T>(one.ThrowOnFailure(...));
    if (_subject.Two() is Option<T> two)
        return AndUnionValueConstraint<T>(two.ThrowOnFailure(...));

    throw new InvalidOperationException(...);
}
```

Justin-Cooney replied:
Yes, that's a good point. The only down-side to this approach is that it is possible for the developer to specify a Type which doesn't exist in the `Union`. I'm not sure if that is of any concern though as it shouldn't be too difficult for the developer to stay on top of which types are a part of each Union. If this is acceptable I can update to your suggestion of a single method with a type checking chain.

I do like the idea of some assertion to ensure that the type are all different. This also raises the question of should we add a method for checking types by position in the `Union` in case the developer has a `Union` with duplicated types?

```C#
public AndUnionValueConstraint<T> BeOfTypeOne<T>(...)
public AndUnionValueConstraint<T> BeOfTypeTwo<T>(...)
...
```

RyanMarcotte replied:
Should we just be defining methods `BeOfTypeOne`, `BeOfTypeTwo`, etc?  I think that would be less awkward than having to explicitly specify the type.  Additionally, there is an already incentive to not change the types and ordering of discriminated unions due to the breaking changes involved.

```
public class UnionValueTypeAssertions<TUnionType, TUnionDefinition, T1, T2>
{
	public AndUnionValueConstraint<T1> BeOfTypeOne(...)
	{
        	Execute.Assertion
			.BecauseOf(because, becauseArgs)
			.ForCondition(_subject.One().HasValue())
			.FailWith(() => ...);

		return new AndUnionValueConstraint<TExpected>(_subject.One().ThrowOnNone(...));
	}

	public AndUnionValueConstraint<T2> BeOfTypeTwo(...) => ...
}
```

Justin-Cooney replied:
This is a good point to think about. I've implemented this methods and given them a bit of a test. 

Implementation-wise they are much cleaner. I definitely like not having to do the ignore params hack or allowing users to enter invalid types.

On the cons side I have noticed that it may be bit difficult to tell which position of the union you are looking for when using a union definition. Hovering over a union using a definition will display `Union<SomeFailure>`, not the actual list of types so you may find yourself trying each position to find the type you are looking for. It is also slightly less clear what type is being asserted at a glance.

That's mostly speculation on my part though, I'm all for going with the method per position route and seeing how that works in practice.

RyanMarcotte replied:
Depending on how many assertions you have to write for a particular union in a test suite, it may be worth providing some custom `BeYourUnionPartType` extension method for that union type (a `YourUnionTypeAssertions` class) that delegates to the `BeOfTypeOne` / `BeOfTypeTwo` extension method underneath.  I realize that the point of these extension methods is to help get rid of boilerplate code, but the "specification" nature of `Functional.Unions` means you'll end up having to write custom code to get the syntax / readability result you want if `BeOfTypeOne`, `BeOfTypeTwo` is not what you want the test code to look like.

Without the extra code, Intellisense should help with the type discoverability, no?  Given `union.Should().`, would `BeOfTypeOne` and `BeOfTypeTwo` not show up as selections?  If so, that helps the writability part at least.

Justin-Cooney replied:
Good observations, that makes sense to me. The intellisense does offer some assistance in discoverability and additional extensions are always an option.

I'll be removing the old methods and leaving just `BeOfTypeOne / BeOfTypeTwo` before I create assertions for all `Union` sizes. Thanks for the feedback!

RyanMarcotte commented on src/Functional.Unions.FluentAssertions/Functional.Unions.FluentAssertions/Extensions/OptionExtensions.cs:
```diff
@@ -0,0 +1,10 @@
+{
+	internal static class OptionExtensions
+	{
+		public static T ValueUnsafe<T>(this Option<T> source)
```
The latest `Functional.Primitives.Extensions` package [defines a `ThrowOnNone` extension method](https://github.com/JohannesMoersch/Functional/pull/106) (2.4.0)

Justin-Cooney replied:
Nice addition! I will use that instead.

RyanMarcotte commented:
Once we have solidified the API, please add documentation for the `Functional.Union` assertions in the main `README.md` file.

RyanMarcotte commented:
I am going to rename the repo to `Functional.FluentAssertions`.  The main `README` can then point to a `Functional.Primitives.FluentAssertions` page and a `Functional.Unions.FluentAssertions` page.

RyanMarcotte commented:
Once we have solidified the API, are you going to be adding `Task`-based assertion methods for `Functional.Union` types as part of your changes as well?  It may make more sense to include that work with #9 after this work gets merged in.

Justin-Cooney commented:
> Once we have solidified the API, are you going to be adding `Task`-based assertion methods for `Functional.Union` types as part of your changes as well? It may make more sense to include that work with #9 after this work gets merged in.

Good thinking. I'll restrict this PR purely to the Unions themselves and then add `Task` extensions are part of #9

RyanMarcotte commented on src/Functional.Unions.FluentAssertions/Functional.Unions.FluentAssertions.Tests/UnionValueTypeAssertionsTests.cs:
```diff
@@ -0,0 +1,72 @@
+{
+	public class UnionValueTypeAssertionsTests
+	{
+		public class TOne { }
```
To reduce potential confusion with the naming convention associated with generic parameters, rename `TOne` and `TTwo` classes to something else?

``` csharp
// example...  feel free to define your own
public class LotTracked
{
    public string LotCode { get; }
}

public class NotTracked
{
}
```

Justin-Cooney replied:
Good suggestion, I could see that getting confusing. Updated in the latest commit!

Justin-Cooney committed:
Renames union test classes.

Justin-Cooney committed:
Updates Union library to use `ThrowOnNone`. Updates FunctionalUnionAssertions namespace to `Functional`.

Justin-Cooney committed:
Adds BeOfTypeOne and BeOfTypeTwo Union Assertions

Justin-Cooney committed:
Removes old BeOfType Union assertions and cleans up unit tests.

Justin-Cooney commented:
Update the `Union<TOne, TTwo>` assertions to use `BeOfTypeOne / BeOfTypeTwo`. If everything looks good at this point I will do the following:
- Add assertions for Unions of size 3-8
- Add ReadMe documentation
- Update project package info

RyanMarcotte commented:
![image](https://user-images.githubusercontent.com/1458084/95129780-bdcefa80-0718-11eb-9800-3014b0461a1c.png)

Justin-Cooney committed:
Adds assertion extensions for Unions with three or more types.

Justin-Cooney committed:
Adds Readme for `Functional.Unions.FluentAssertions`.

Justin-Cooney committed:
Initializes package information for the `Functional.Unions.FluentAssertions` project.

Justin-Cooney committed:
Updates `Functional.Unions.FluentAssertions` assembly info to v3.0.0

RyanMarcotte commented on src/Functional.Unions.FluentAssertions/Functional.Unions.FluentAssertions.Tests/UnionValueTypeAssertionsTests.cs:
```diff
@@ -0,0 +1,1266 @@
+		public class SevenDefinition : UnionDefinition<SevenDefinition, ClassOne, ClassTwo, ClassThree, ClassFour, ClassFive, ClassSix, ClassSeven> { }
+		public class EightDefinition : UnionDefinition<EightDefinition, ClassOne, ClassTwo, ClassThree, ClassFour, ClassFive, ClassSix, ClassSeven, ClassEight> { }
+
+		public class AdHocWithTwoTypes
```
Can we break this file into multiple files?  Resharper (Ctrl+R,O) makes it easy.
`UnionValueTypeAssertionsTests.AdHocWithTwoTypes.cs`
`UnionValueTypeAssertionsTests.DefinitionWithTwoTypes.cs`
... etc

Justin-Cooney replied:
Sure thing, multiple files always help!

Justin-Cooney commented:
Completed and added Readme + package info.

I've added a separate ReadMe for unions as I believe the plan was to split into separate ones and link from the root ReadMe.
Also, I've initialized the package info as v3.0.0

RyanMarcotte commented on src/Functional.Unions.FluentAssertions/Functional.Unions.FluentAssertions/Functional.Unions.FluentAssertions.csproj:
```diff
@@ -0,0 +1,22 @@
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <Authors>Ryan Marcotte + contributors</Authors>
+    <Version>2.1.0</Version>
```
Should be version `1.0.0`, no?

Justin-Cooney replied:
Oh yes, I think I was a little confused because I wasn't sure if the two packages had to match version numbers. I suppose other libraries like `Functional` only have sub packages match version numbers because there is an encompassing `Functional.All` package.

RyanMarcotte replied:
Correct.  The `Functional` library is set up such that a version bump affects all packages.  These packages are independent, so their version numbers will vary independently.

RyanMarcotte replied:
With that in mind, you can totally slap your name in the `Authors` space instead of mine 🙂

RyanMarcotte commented on src/Functional.Unions.FluentAssertions/Functional.Unions.FluentAssertions/UnionValueTypeAssertions.cs:
```diff
@@ -0,0 +1,772 @@
+
+namespace Functional.Unions.FluentAssertions
+{
+	public class UnionValueTypeAssertions<TUnionType, TUnionDefinition, TOne, TTwo>
```
Split this into multiple files as well?  I'm OK with `UnionValueTypeAssertions2.cs`, `UnionValueTypeAssertions3.cs`, etc.

Justin-Cooney replied:
Will do, thanks for the suggestions!

RyanMarcotte reviewed (changes requested):
I assume that the `Task<Union<...>>` stuff will be written after everything else checks out?

RyanMarcotte commented:
> I've added a separate ReadMe for unions as I believe the plan was to split into separate ones and link from the root ReadMe.

Yep, the existing main `README.md` file can be moved into the `Functional.Primitives.FluentAssertions` folder and a new `README.md` file in the repository's root can direct users to the two pages.

Justin-Cooney committed:
Splits Union assertions and test classes into multiple files.

Justin-Cooney committed:
Changes Union assembly version to v1.0.0

Justin-Cooney committed:
Merge branch 'master' into UnionAssertionsFeature

Justin-Cooney committed:
Merges changes from master into UnionAssertionsFeature

RyanMarcotte commented:
Are the `Task<Union<...>>` assertions being added as part of this PR too?

Justin-Cooney commented:
> Are the `Task<Union<...>>` assertions being added as part of this PR too?

Seeing as the other `Task` extensions are already in master I think I will take the time to add those to this PR.

Justin-Cooney committed:
Splits the library readmes and links to them from a common readme.

Justin-Cooney committed:
Merge branch 'UnionAssertionsFeature' of https://github.com/RyanMarcotte/Functional.Primitives.FluentAssertions into UnionAssertionsFeature

Justin-Cooney committed:
Adds Task wrapped Union assertions.

Justin-Cooney committed:
Updates the `Functional.Unions.FluentAssertions` README with information on Task assertions.

Justin-Cooney commented:
Latest commits:
- Split the READMEs
- Updates to package info
- Division into files per union size
- Addition of `IUnionTask` extensions
- README update for `IUnionTask` extensions

RyanMarcotte committed:
in-code documentation

ChrisDusyk commented:
I'm just checking this all out now, but this is extremely exciting to see in here. Thanks @Justin-Cooney and @RyanMarcotte!!
